### PR TITLE
Move to specify shots on the QNode

### PR DESCRIPTION
**Pull request by albi3ro**
We are moving to define shots as part of **what** we execute, not necessarily **how** we execute it.

We already:
* Allow shots to be overridden on each call in the `QNode`
* Allow `override_shots` to be provided to `qml.execute`
* Allow shots to be specified in the `QuantumTape`
* Purely take shots from the `QuantumTape` in the new device API.

These changes will allow us to have better shot distribution within batches, since each tape in a batch will be able to have a separate shot number.

This will also help PennyLane become more functional, since we will not be needing to constantly mutate the device in order to support per-call dynamic shots.

Since these warnings are inside the `QNode` and the `qml.device` constructor, no changes are currently necessary for plugins other than updated tests and documentation.

## Example behaviour for deprecation cycle

```python
import pennylane as qml
import warnings
warnings.simplefilter("always")
```

Don't specify shots on device? No warning

```python
qml.device('braket.local.qubit', wires=2)
```
    <BraketLocalQubitDevice device (wires=2, shots=None) at 0x14001c400>

Manually specify shots on device? Warning

```python
qml.device('braket.local.qubit', wires=2, shots=None)
```
    /Users/christina/Prog/pennylane/pennylane/__init__.py:309: UserWarning: In v0.33, the shots will always be determined by the QNode. Please specify shots there.
      warnings.warn("In v0.33, the shots will always be determined by the QNode. Please specify shots there.", UserWarning)
    <BraketLocalQubitDevice device (wires=2, shots=None) at 0x17fbd8a00>

Don't specify shots on either device or qnode? No warning

```python
dev = qml.device('braket.local.qubit', wires=2)

@qml.qnode(dev)
def circuit():
    return qml.expval(qml.PauliX(0))
```

Specify shots on device but not qnode? 

warning, and make qnode shots equal to device shots

```python
dev = qml.device('braket.local.qubit', wires=2, shots=10)
```
    /Users/christina/Prog/pennylane/pennylane/__init__.py:309: UserWarning: In v0.33, the shots will always be determined by the QNode. Please specify shots there.
      warnings.warn("In v0.33, the shots will always be determined by the QNode. Please specify shots there.", UserWarning)

```python
@qml.qnode(dev)
def circuit():
    return qml.expval(qml.PauliX(0))

circuit(), circuit.default_shots
```
    /Users/christina/Prog/pennylane/pennylane/qnode.py:448: UserWarning: Shots should now be specified on the qnode instead of on the device.Using shots from the device. QNode specified shots will be used in v0.33.
      warnings.warn(

    (tensor(0.2, requires_grad=True),
     Shots(total_shots=10, shot_vector=(ShotCopies(10 shots x 1),)))

Shots not none on qnode? no warning when creating the qnode, but there is a warning when putting on them on the device still.

```python
dev = qml.device('lightning.qubit', wires=2, shots=1000)
```
    /Users/christina/Prog/pennylane/pennylane/__init__.py:309: UserWarning: In v0.33, the shots will always be determined by the QNode. Please specify shots there.
      warnings.warn("In v0.33, the shots will always be determined by the QNode. Please specify shots there.", UserWarning)

```python
@qml.qnode(dev, shots=2)
def circuit():
    return qml.sample(wires=0)

circuit()
```
    tensor([0, 0], requires_grad=True)

### CODE REVIEW

**Comment by albi3ro** [comment]
[sc-41930]

File: pennylane/qnode.py [hunk 1]
```diff
@@ -443,13 +444,26 @@ def __init__(
...
             )
 
+        if (
+            hasattr(device, "shots")
+            and device.shots != shots
+            and device.shots is not None
+            and shots is None
+        ):
+            warnings.warn(
+                "Shots should now be specified on the qnode instead of on the device."
```
Comment (trbromley) [inline pennylane/qnode.py:454]:
```suggestion
                "The shots value you provided to the device does not match the value provided to the QNode."
```

File: pennylane/qnode.py [hunk 2]
```diff
@@ -443,13 +444,26 @@ def __init__(
...
+                "Using shots from the device. QNode specified shots will be used in v0.33."
```
Comment (trbromley) [inline pennylane/qnode.py:455]:
```suggestion
                "Using shots from the device. QNode specified shots will be used in v0.33 "
                "and shots will no longer be permitted as a device argument."
```

**Comment by albi3ro** [comment]
Closed in favor of #4388

**Pull request closed without merging**